### ISSUE TITLE
Mux.js Usage of .concat questionable

### ISSUE THREAD

**Issue by Uzlopak** [amqp-node/amqplib#657]
I think we could improve the performance and memory footprint of Mux.js if we  change
https://github.com/squaremo/amqp.node/blob/243aef8b6da766a3b9908bc50b2be4859eef43a6/lib/mux.js#L82

to this:
```javascript
Array.prototype.push.apply(this.oldStreams, this.newStreams);
```

or 
```javascript
this.oldStreams.push(...this.newStreams);
```

.concat is creating a new Array. But if we directly push the elements of newStreams-Array directly into the oldStreams-Array, we would actually reduce the memory overhead and probably would gain some minor performance gains.

**Comment by carlhoerberg** [comment]
If you're looking for performance i would recommend https://github.com/cloudamqp/amqp-client.js/. We've optimized it quite heavily and are open to more improvements.

**Comment by Uzlopak** [comment]
@carlhoerberg 

Do you have benchmarks for your claim :)?

**Comment by carlhoerberg** [comment]
Yes, readme updated

**Comment by Uzlopak** [comment]
Very interesting. Thank you.

**Comment by Uzlopak** [comment]
@carlhoerberg 

Can you supply the benchmarks also into the repo? Maybe something, where you have to spawn  a rabbitmq-server via docker-compose?

**Comment by cressie176** [comment]
Switching to from concat to Array.prototype.push.apply makes a substantial difference to performance...

```javascript
const assert = require('assert');

class Appender {
  constructor(size) {
    this.size = size;
    this.a = [];
    this.b = new Array(size).fill().map(Math.random);
  }
}

class Push extends Appender {
  test() {
    Array.prototype.push.apply(this.a, this.b);
  }
}

class Concat extends Appender {
  test() {
    this.a = this.a.concat(this.b);
  }
}

function test(variant, iterations) {
  const before = Date.now();
  for (let i = 0; i < iterations; i++) {
    variant.test();
  }
  const after = Date.now();
  assert(variant.a.length, variant.size * iterations)
  return after - before;
}

function testMany(variant, iterations) {
  return new Array(iterations)
    .fill()
    .map(() => test(variant, iterations))
    .reduce((total, duration) => (total + duration), 0);
}

const size = 10;
const iterations = 100;

const push = testMany(new Push(size), iterations);
const concat = testMany(new Concat(size), iterations);
console.log({ push, concat })
```

```
{ push: 4, concat: 1964 }
```
I tried with other combinations of `size` and `iterations`, and other than for very small numbers there was a substantial improvement using push.

**Comment by cressie176** [comment]
I've merged into main and will close this issue with the next release

**Comment by cressie176** [comment]
published in amqplib@0.9.1 Thank you @Uzlopak
